Introduce Issue-Extension with Service-API (#135 / #361)

Bukama opened the pull request:
This PR shows an implementation of an issue extension , which publishes the results of annotated tests through an API.

```
Introduce Issue-Extension with Service-API (#135 / #361)

This PR provides the @Issue-Annotation and its extension,
which allows users to show that tests belong to a special
issue (e.g. requirement).

While this information is not needed during test execution
(e.g. like changing the time zone via @DefaultTimeZone-
annotation), but after the test execution (e.g. in a report
to check which tests of which issues have passed/failed),
this PR also introduces the IssueExtensionReportListener.

This listener collects the name, issueid and result of all
tests which are annotated with the introduced annotation.
After the test execution has finished, the collected
information are published to the IssueProcessor API.

Users can implement this API for further process of the
results based on their individual needs, e.g. generating a
report, or sending mails.

With such an API Pioneer does not makes any assumptions
about the users needs and also avoids additional
dependencies and therefore dependency conflicts.

closes #135
PR: #361
```

closes #135

Michael1993 reviewed (changes requested):
I really like this approach, it looks very good.

Relying on the "Issue" string as a report entry key is really restrictive, other than that all my other comments are minor things.

Bukama commented on src/main/java/org/junitpioneer/jupiter/IssueExtension.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @see Issue
+ */
+class IssueExtension implements BeforeEachCallback {
```
I don't think so, because (when I understand it correctly) the `TestWatcher` has a callback for each of the result, but the extension only needs to publish the annotations value.

Bukama commented on src/main/java/org/junitpioneer/jupiter/IssueExtension.java:
```diff
@@ -0,0 +1,49 @@
+	 * @param context The Extensions context
+	 * @return The read value
+	 */
+	String readIssueIdFromAnnotation(ExtensionContext context) {
```
Same question, with same answer as in every PR.

Bukama commented on src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionListener.java:
```diff
@@ -0,0 +1,87 @@
+ */
+public class IssueExtensionListener implements TestExecutionListener {
+
+	static final String KEY_ISSUE = "Issue";
```
I could change it to `org.junit-pioneer.issue`, but then you still argue, that people want to publish other report entries with this key. Same with every generated value, like an UUID.

Bukama commented on src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionListener.java:
```diff
@@ -0,0 +1,87 @@
+	private final ConcurrentHashMap<String, String> testStatusCache = new ConcurrentHashMap<>();
+
+	// Package private by purpose for testing
+	List<IssuedTestCase> allIssuedTests = new ArrayList<>();
```
Then I would add code just for testing in a productive code base, which is bad coding. Package private variables, methods, etc. are best practices for testable code.

I know that you don't like that (see your comment above about package private method), but I keep this coding style fully minded. I had this discussions for years at my work. I tested my methods and had almost no errors in my code, while colleagues had uncountable amount of errors, because they uses private methods and then claimed, that they could not test their code.

Bukama commented on src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionListener.java:
```diff
@@ -0,0 +1,87 @@
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		if (testIdentifier.isTest()) {
+			// Store test result in cache
+			testStatusCache.put(testIdentifier.getUniqueId(), testExecutionResult.getStatus().toString());
```
The `TestExecutionResult` only contains the status and maybe a `Throwable`, but the extension only cares about the status.

Bukama commented on src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionListener.java:
```diff
@@ -0,0 +1,87 @@
+			allTests.forEach(testID -> {
+				String status = testStatusCache.getOrDefault(testID, "UNKNOWN");
+
+				allIssuedTests.add(new IssuedTestCase(testID, issueId, status));
```
Because I combine the entries of two different `Map`s into a single `List`.

Bukama commented on src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionListener.java:
```diff
@@ -0,0 +1,87 @@
+		}
+
+		// Pass results to all IssueProcessors
+		Iterator<IssueProcessor> processors = IssueProcessorProvider.getInstance().providers();
```
The `ServiceLoader` tutorial I worked with showed it that way, to encapsulate the retrieval of the `ServiceLoader`.

Bukama replied:
Changed

Bukama commented on src/main/java/org/junitpioneer/jupiter/issue/IssueProcessor.java:
```diff
@@ -0,0 +1,27 @@
+	 *
+	 * @param allResults List of tests with their unique name, issueId and execution results.
+	 */
+	void processTestResults(List<IssuedTestCase> allResults);
```
Why make it more complex? All information are in the `IssuedTestCase` and a `List` is easier than a `Map` I think.

nipafx replied:
I'd prefer to pass on more structured information. What about a list of these:

```java
public record TestedIssue(String issueId, List<IssueTestCase>) { }
public record IssueTestCase(String testId, Status result) { }
```

But obviously not as records. 😊

nipafx replied:
I also think that gathering these information in separate maps is a bit iffy. I prefer richer data structures - maybe a `Map<TestId, TestedIssueBuilder>`?

Bukama replied:
> I'd prefer to pass on more structured information. What about a list of these:
> ```
> public record TestedIssue(String issueId, List<IssueTestCase>) { }
> public record IssueTestCase(String testId, Status result) { }
> ```
Changed

> I also think that gathering these information in separate maps is a bit iffy. I prefer richer data structures - maybe a `Map<TestId, TestedIssueBuilder>`?

I don't understand what you mean with TestedIssueBuilder

Bukama commented on src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorProvider.java:
```diff
@@ -0,0 +1,50 @@
+	 *
+	 * @return Instance of IssueProcessorProvider
+	 */
+	public static IssueProcessorProvider getInstance() {
```
The tutorial I worked with showed it that way.

Bukama replied:
Changed

Bukama commented on src/main/java/org/junitpioneer/jupiter/issue/IssuedTestCase.java:
```diff
@@ -0,0 +1,68 @@
+
+	private final String uniqueName;
+	private final String issueId;
+	private final String result;
```
I don't agree :D

nipafx replied:
I wanted to recommend `TestExecutionResult` as well, but it doesn't know the test's id, so we have to store that ourselves anyway. Instead of referencing it, we can just repack the information we need into our own class - see my proposal for `IssueTestCase` above.

Bukama commented on src/main/resources/META-INF/services/org.junit.platform.launcher.TestExecutionListener:
```diff
@@ -0,0 +1 @@
+org.junitpioneer.jupiter.issue.IssueExtensionListener
```
I have to read that up when I find some time.

nipafx commented on src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionListener.java:
```diff
@@ -0,0 +1,87 @@
+
+	static final String KEY_ISSUE = "Issue";
+
+	// Storage with all tests that belong to an issue <issueId, List<UniqueIdentifier>>
+	private final ConcurrentHashMap<String, List<String>> issueTestsStorage = new ConcurrentHashMap<>();
+
+	// Storage with tests results of test cases <UniqueIdentifier, result>
+	private final ConcurrentHashMap<String, String> testStatusStorage = new ConcurrentHashMap<>();
```
This extension isn't stateless. As far as I'm aware, Jupiter doesn't guarantee it will use the same instance for the entire test run (although I couldn't verify that in the user guide). I wanted to propose to use the store to store and retrieve all necessary information, but I just see that `TestExecutionListener`s don't have access to that (makes sense, because `TestExecutionListener` doesn't know anything about Jupiter, which is where the store lives).

nipafx commented on src/main/java/org/junitpioneer/jupiter/IssueExtension.java:
```diff
@@ -0,0 +1,49 @@
+ */
+class IssueExtension implements BeforeEachCallback {
+
+	static final String KEY = "Issue";
```
Yes, I removed the duplication.

nipafx commented on build.gradle.kts:
```diff
@@ -42,9 +42,9 @@ dependencies {
 	implementation(group = "org.junit.jupiter", name = "junit-jupiter-params")
 	implementation(group = "org.junit.platform", name = "junit-platform-commons")
+	implementation(group = "org.junit.platform", name = "junit-platform-launcher")
 
 	testImplementation(group = "org.junit.jupiter", name = "junit-jupiter-engine")
-	testImplementation(group = "org.junit.platform", name = "junit-platform-launcher")
```
sormuras popped into the stream and answered this with a riddle three before eventually letting us know that there's no problem with this.

nipafx commented on src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorProvider.java:
```diff
@@ -0,0 +1,50 @@
+	 *
+	 * @return All known {@link IssueProcessor}s
+	 */
+	public Iterator<IssueProcessor> providers() {
+		return loader.iterator();
```
I don't see the use for this class either. It just seems to be a wrapper around `ServiceLoader.load(IssueProcessor.class)`. No reason not to call that directly wherever you need it.

nipafx commented on src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionListener.java:
```diff
@@ -0,0 +1,87 @@
+ * <p>This listener collects the names and results of all tests, which are annotated with the {@link org.junitpioneer.jupiter.Issue} annotation.
+ * After all tests are finished the results are provided to an {@link IssueProcessor} for further processing.</p>
+ */
+public class IssueExtensionListener implements TestExecutionListener {
```
I think we should include `Execution` in the name, e.g. `IssueExtensionExecutionListener`.

nipafx commented on src/main/module/module-info.java:
```diff
@@ -2,6 +2,7 @@
 	requires org.junit.jupiter.params;
 	requires org.junit.platform.commons;
+	requires org.junit.platform.launcher;
 
```
This needs to be here for services to work on the module path:

```
	provides org.junit.platform.launcher.TestExecutionListener
			with org.junitpioneer.jupiter.issue.IssueExtensionListener;
```

(I wanted to propose a change, but I can't propose one at the end of the file, which is where this should go.)

nipafx reviewed (commented):
Thank you for the contribution, @Bukama! 👍🏾 I made a few small changes (63630f5e) and left some comments below. Two larger topics remain to be discussed:

## Method scope

Don't you think it would make sense to annotate an entire test class (or even package) that implements a feature with the annotation instead of having to repeat it on every test? We could kick this down the road, though, and publish with method-scope and open a bishue for further improvements.

## Package structure

Assuming we remove `IssueProcessorProvider`, we have the following classes:

* `@Issue` - needs to be accessible
* `IssuedTestCase` or whatever we replace it with - need to be accessible
* `IssueProcessor` - needs to be accessible
* `IssueExtensionListener` - needs to be public (for the `ServiceLoader` to pick it up), but inaccessible (not part of our API)

I think it would be best to put only `IssueExtensionListener` this `org.junitpioneer.jupiter.issue` and everything else into `org.junitpioneer.jupiter` and then we don't add any export. If we go this route, my proposal for the class `TestedIssue` should be reconsidered - having the name start with `Issue` would be nicer as then all needed classes are in the same place and start with the same phrase.

Bukama commented:
> Two larger topics remain to be discussed:
> 
> ## Method scope
> Don't you think it would make sense to annotate an entire test class (or even package) that implements a feature with the annotation instead of having to repeat it on every test? We could kick this down the road, though, and publish with method-scope and open a bishue for further improvements.

Added possibility for class level.

> 
> ## Package structure
> Assuming we remove `IssueProcessorProvider`, we have the following classes:
> 
> * `@Issue` - needs to be accessible
> * `IssuedTestCase` or whatever we replace it with - need to be accessible
> * `IssueProcessor` - needs to be accessible
> * `IssueExtensionListener` - needs to be public (for the `ServiceLoader` to pick it up), but inaccessible (not part of our API)
> 
> I think it would be best to put only `IssueExtensionListener` this `org.junitpioneer.jupiter.issue` and everything else into `org.junitpioneer.jupiter` and then we don't add any export. If we go this route, my proposal for the class `TestedIssue` should be reconsidered - having the name start with `Issue` would be nicer as then all needed classes are in the same place and start with the same phrase.

Classes moved and renamed

nipafx commented:
Screw Sonar and its insistence on test coverage. MERGE AWAY!
